First draft of groups overview

viv-liu opened the pull request:
A bunch of scribbles at the end

Commit:
Add files via upload

JimacoMS2 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+# Working with groups in Microsoft Graph
```
So just a general comment, I think we need follow the format for overview topics that Angela and Laura have been working on. Good example is here:  https://developer.microsoft.com/en-us/graph/docs/api-reference/v1.0/resources/mail_api_overview with the addition of an Auth section similar to what is in the AAD overview topics: https://developer.microsoft.com/en-us/graph/docs/api-reference/v1.0/resources/azure_ad_overview#authorization .

JimacoMS2 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+- add and remove owners of a group.
+
+**NOTE**
+Groups are only supported for enterprise accounts, and not supported for Microsoft personal accounts.
```
check with Laura about using "enterprise" instead of "work or school"

dkershaw10 replied:
We should adhere to our glossary :)

JimacoMS2 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+# Working with groups in Microsoft Graph
+
+Groups are collections of [users](*) and [devices](*) (and more?) who share access to resources in Office365 or in your app. All operations in Microsoft Graph on groups require administrator consent. Using Microsoft Graph, you can perform the following operations and more:
```
So I'm not quite sure about this opening statement. AFAIK Office 365 groups (all flavors) can only have users as members, Security groups can have users and groups, Outlook distribution groups can have users, groups and contacts. Not sure about devices for any of these and a solid emphasis on "AFAIK"  :). Also, O365 groups are all about collaboration, while security groups are about controlling access to resources in a directory or app(?).

dkershaw10 replied:
Agreed - and this is part of the challenge because we have many group types.  I would open with the traditional view of groups and then extend.  Something like:  "Traditionally, groups are logical containers or collections of users, that are most commonly used to control access to resources - whether this is to Microsoft services or to your own app.  Microsoft Graph provides you with APIs to create and manage a variety of different types of groups and group functionality to suit your scenario needs."
Somehow we need to continue this to really draw the reader to use Office Groups, and when it's appropriate to use the other group types.  We really need to make this definitive guidance IMHO

viv-liu replied:
Agreed. I'll try adding in a "Common scenarios" table that is more helpful instead of this flat list.

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+# Working with groups in Microsoft Graph
+
+Groups are collections of [users](*) and [devices](*) (and more?) who share access to resources in Office365 or in your app. All operations in Microsoft Graph on groups require administrator consent. Using Microsoft Graph, you can perform the following operations and more:
+- create groups
```
Maybe these are "core" group operations?  Not sure if we should add update and delete here (to create).

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+Groups are only supported for enterprise accounts, and not supported for Microsoft personal accounts.
+
+## Office 365 groups
+Office 365 groups are for people who collaborate on a project or a team. Members share access to Office365 resources including:
```
You should take a look at this doc:  https://support.office.com/en-us/article/Learn-about-Office-365-groups-b565caa1-5c40-40ef-9915-60fdb2d97fa2 and pull some info from it, and also link to it.

viv-liu replied:
Added a link and pulled some content.

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+- calendar
+- drive
+- notes
+- plans (beta)
```
I thought this was GA?  I see it in the metadata in v1.0 (planner).

viv-liu replied:
oops my bad

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+
+| Type | Use case | Shared resources |
+|------------|-------|--------|
+| Outlook group | Mail-driven collaboration | Mail conversations and threads, tasks (beta) |
```
Are we sure outlook tasks are available in groups?  I think this is only available on /users or /me.
Also is this the technical term according to Mike and Kairaz - Outlook Group and Yammer Group, or should it be:

Microsoft Team, Yammer-based conversations, Outlook-based conversations?

Also "Outlook-based conversation" is the default when creating an Office Group through the API.  Given that's all you can do right now, not sure how we should put that...

viv-liu replied:
Tasks is not available for groups (at least how I tried it). I will verify with the other two for the terminology.

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+| Type | Use case | Shared resources |
+|------------|-------|--------|
+| Outlook group | Mail-driven collaboration | Mail conversations and threads, tasks (beta) |
+| Microsoft Team | High velocity chat-driven collaboration | Channels, chat threads |
```
Is this what Teams calls it?  They normally refer to it as "persistent chat". See https://developer.microsoft.com/en-us/graph/docs/api-reference/beta/resources/teams_api_overview

viv-liu replied:
Added persistent

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+| Microsoft Team | High velocity chat-driven collaboration | Channels, chat threads |
+| *TODO* Yammer group | How to access Yammer posts through API? |
+
+**NOTE** Today, only Outlook groups can be created through the API. 
```
"Currently" rather than "Today"

viv-liu replied:
Done

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+
+**NOTE** Today, you can check if an Office 365 group is an Outlook group or a Microsoft Teams by trying GET /group/{id}/channels. If it works, then it's a Microsoft Team, else it's an Outlook group.
+
+### Example of Outlook group
```
OK - there's nothing in here about what are the properties I need to set to create an Office Group (GroupType, mailEnabled = true, mailNickName, mail(?), and securityEnabled=false.

viv-liu replied:
I think the details for creating should be in the create group article. I'll add something to that and you can tell me how you like it.

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
````diff
@@ -0,0 +1,273 @@
+}
+```
+## Security group and mail-enabled security group
+Security groups are for controlling user access to resources. By checking whether or not a user is a member of a security group, your app can make authorization decisions when that user is trying to access some secure resources in your app. Mail-enabled security groups are the same on principle, but with a group mailbox provisioned by default so the group can receive mail.
````
same **in** principle
EXCEPT - I don't think you can create mail-enabled SGs through MS Graph - or can you?

viv-liu replied:
Good question. I tested this, couldn't create a MESG through the API

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+## Security group and mail-enabled security group
+Security groups are for controlling user access to resources. By checking whether or not a user is a member of a security group, your app can make authorization decisions when that user is trying to access some secure resources in your app. Mail-enabled security groups are the same on principle, but with a group mailbox provisioned by default so the group can receive mail.
+
+**OPENQ** Can SGs be nested?
```
Yes.

viv-liu replied:
Addressed

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+
+**OPENQ** Can SGs be nested?
+
+### Example of security group
```
A number of these properties are ONLY relevant to Office Groups...

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+**NOTE**
+Groups are only supported for enterprise accounts, and not supported for Microsoft personal accounts.
+
+## Office 365 groups
```
What licenses do you need for Office Groups?

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+**NOTE** Dynamic membership rules requires the tenant to have a license at tier [Azure Active Directory Premium P1](https://azure.microsoft.com/en-us/pricing/details/active-directory/) or greater.
+
+## Administrator master switch
+Where can an admin disable groups master switch. Planner today will show error. 
```
Needs more.  You can reference the new docs on GroupSettings.

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+
+# END OF ARTICLE, the rest are notes and scribbles
+
+## Distribution group (not sure if this is possible through MS Graph, talk to Dhana)
```
We should be pushing folks to using Office Groups.  DGs should really be "legacy".  At some point we should switch off DG creation IMHO.  You should look for messaging guidance from Mike and Kairaz.  They really should write this section.

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,273 @@
+    "securityEnabled": false,
+}
+
+{
```
This is an awesome error.  You should file a bug for this.

Commit:
V1.5 still needs a common scenarios type table

Addressed most of the comments

Commit:
Added common use cases table

This is an area of **general** functionality. More specific use cases related to Outlook groups and Teams groups should be ideally filled in by those who have a better understanding of what developers want to do with those types of groups. 

Also updated the examples to be more generic.

Commit:
Added differentiating table and minor fixes

Table that differentiates O365 and security groups. 

Yet to add groupSettings to this...

Commit:
Incorporated feedback, added other types table

In this article, only O365 groups - Outlook flavor and SGs are described. Teams is not mentioned since it has its own node in /beta docs. At the bottom of this doc are other types of groups that can't be managed through the API, but can be GET.

JimacoMS2 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,126 @@
+- SharePoint Team Site
+- Planner
+
+Your app can access and manage these resources through the API.
```
I'd like to see some discussion of the difference between Outlook groups and Microsoft Teams -- including the different resources they support. (maybe a table :) ). I think of the journey I had to go on to understand this and the fact that I had internal resources to get me there and I think we are doing our customers a disservice if we don't make this clear.

dkershaw10 replied:
I wonder if there's more information that we can provide here - and more scenarios that really help sell the value-proposition of groups to a developer.
Kairaz and Mike need to comment on what they'd like to see here.

When creating a "Unified" group, the default group that is created is an Office 365 Group.  Other variants are possible and built on top of the Office 365 Group concept.  For example Microsoft Teams contains all the feature of Office 365 groups and more.  However creation of a Microsoft Team through the Microsoft Graph API is not currently available. For more on Microsoft Teams, see ... (link to their landing page).

viv-liu replied:
The difficulty here is that Microsoft teams is its own thing, separate from groups right now in the TOC. We can certainly describe them here, but wouldn't a developer just go click the Microsoft Teams overview? The two overviews should fit together.

JimacoMS2 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,126 @@
+Learn more about Office 365 groups and the administrator experiences [here](https://support.office.com/en-us/article/Learn-about-Office-365-groups-b565caa1-5c40-40ef-9915-60fdb2d97fa2).
+
+## Security groups
+Security groups are for controlling user access to resources. By checking whether or not a user is a member of a security group, your app can make authorization decisions when that user is trying to access some secure resources in your app. Security groups can have users and other security groups as members.
```
Again -- please vet the focus on controlling in-app resources with Dan K.

JimacoMS2 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,126 @@
+## Security groups
+Security groups are for controlling user access to resources. By checking whether or not a user is a member of a security group, your app can make authorization decisions when that user is trying to access some secure resources in your app. Security groups can have users and other security groups as members.
+
+### Example of security group
```
If we have vert space for the examples great -- if not, we can remove them.

viv-liu replied:
I think the example is nice, and I think so far we have the vert space. The common use cases though, that may get bigger.

JimacoMS2 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,126 @@
+
+Your app can access and manage these resources through the API.
+
+### Example of Outlook group
```
If we have vert space for the examples great -- if not, we can remove them ...

dkershaw10 replied:
You need to change the title here.

viv-liu replied:
Done. Changed title to Example of Office 365 group

JimacoMS2 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,126 @@
+| Check if a user is a member of a group, get all the groups the user is a member of. | [user](user.md) <br/> [group](group.md)| [Check member groups](../api/group_checkmembergroups.md) <br/> [Get member groups](../api/group_get_membergroups.md)|
+| List the owners of a group, and add or remove owners. | [user](user.md) <br/> [group](group.md)| [List owners](../api/group_list_members.md) <br/> [Add member](../api/group_post_members.md) <br/> [Remove member](../api/group_delete_members.md)|
+
+## Other types of groups
```
I would actually vote for moving this section above use-cases

viv-liu replied:
Makes sense, done!

JimacoMS2 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,126 @@
+
+| **Use cases**		   | **REST resources**	| **See also** |
+|:---------------|:--------|:----------|
+| **Group object and methods** | | |
```
I would actually like to see this table broken down further so that it could tell me what kinds of things are supported by specific kinds of groups: For example, if create a group had its own row, you could note that only Outlook and pure security groups can be created. If Add members had its own row, you could note that members can only be added for Outlook and security groups, If list members had its own row you could mention that this is supported across all groups, etc. (I'm not saying that every single thing needs its pwn row -- for example, almost all of the reads could probably be grouped together -- just that where there are diff levels of support for different types of groups, they should get their own row.

Also we should break out group functions (that are exposed on directoryObject) and mention the functions that return transitive membership (for security/dist groups anyway) vs. nav properties, like "members" that return direct membership -- note: this does not apply to Office 365 groups (AFAIK they only have direct members). It is important though for security groups and I get questions about it ...

dkershaw10 replied:
+1 on the first paragraph.  Not so much on the second - since I think this is getting into more details than we want.

viv-liu replied:
I agree with Jimaco on expanding the Common use cases table. However, I only know the use cases as listed in articles through the table of contents. I don't actually know about developer scenarios and what developers are looking to do using the API, so I can't really do much other than regurgitate the TOC. Which is pointless. I hope this is an area a scenario owner can further flesh out.

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,126 @@
+# Working with groups in Microsoft Graph
+
+Groups are collections of [users](https://developer.microsoft.com/en-us/graph/docs/api-reference/v1.0/resources/user) who share access to resources in Microsoft services or in your app. Microsoft Graph provides you with APIs to create and manage a variety of different types of groups and group functionality to suit your scenario needs. All operations in Microsoft Graph on groups require administr ...
```
I actually think this intro is OK ( but also not sure about the last sentence).  Maybe to assuage Jimaco, we could say that "Groups are collections of different principals, primarily users (link)."  This makes it broader (and more accurate) for all of our group types, although I worry that "principals" makes it a bit techie.  Then we can qualify it further in the next sentence in terms of how groups are used - i.e. to govern access to resources in Microsoft services or in your apps.  How does that sound?

viv-liu replied:
How about "users and other resources"... I don't know the limitations with different types of groups and the different resources. And I'll remove the last sentence about admin consent.

JimacoMS2 replied:
Maybe just add "and sometimes other principals" to your first sentence? i.e. "Groups are collections of [users](https://developer.microsoft.com/en-us/graph/docs/api-reference/v1.0/resources/user), and, sometimes, other principals, that share access to resources in Microsoft services or in your app. " --- although the other suggestions work too. Only thing about saying "other resources" is that it is a bit open ended -- and we use resources later in the sentence. 

The last sentence about admin consent should be moved to an Authorization section, similar to what we do for AAD overview: https://developer.microsoft.com/en-us/graph/docs/api-reference/v1.0/resources/azure_ad_overview#authorization . For most things you need Group permissions, for some APIs you may also need Directory or User permissions as well (I don't think you need to go into detail.) Group and Directory permissions always require Administrator consent. With delegated permissions, for some operations on groups the signed-in user may need to be a group owner and for others they may need to be a member of an appropriate administrator role. 

Just FYI, I don't think you need to go into  detail in the first sentence, but I think the group breakdown is something like this:
O365 Groups: users (not sure about contacts)
Security groups: users, service principals, other security groups.
Distribution groups: users, contacts, mail-enabled security groups, other distribution groups

The AAD Graph API reference docs are pretty specific about what resources can be members of security and distribution groups.

viv-liu replied:
Added other principals to cover those cases. Also added a new column to capture supported member types

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,126 @@
+| Type              | Use case | groupType | mail-enabled | security-enabled | 
+|-------------------|----------|-----------|--------------|------------------|
+| Office 365 groups | Facilitating user collaboration with shared Microsoft online resources. | ["Unified"] | true | false | 
+| Security groups | Controlling user access to in-app resources. | [] | false | true |
```
Actually Jimaco - that isn't true.  SGs are most commonly/primarily used to restrict/control access to a variety of different services, including LoB enterprise apps, and ISV apps.  We should not call these Azure AD SGs.  Security groups is good.
Vivian - I like the way you've laid this out.  In the first row, I give examples for shared Microsoft online resources (like calendar, conversation, document library etc).

I think (unfortunately) we probably need to mention Mail-enabled security groups and Distribution Groups.

viv-liu replied:
DGs at the bottom since they are not recommended. If you present it as an option at the top then people will try it.

dkershaw10 commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,126 @@
+| Type              | Use case | groupType | mail-enabled | security-enabled | Learn more |
+|-------------------|----------|-----------|--------------|------------------|------------|
+| Groups in Yammer | Facilitating user collaboration with through Yammer posts. | ["Unified"] | true | false | [Yammer developer API docs](https://developer.yammer.com/docs) |
+| Mail-enabled security groups | Controlling user access to in-app resources, with a shared group mailbox. | [] | true | true | [Manage mail-enabled security groups Exchange article](https://technet.microsoft.com/en-us/library/bb123521(v=exchg.160).aspx) |
```
Yes - manage the API is a very broad term.  We need to be careful here.  These 2 group types should be pretty fully manageable through Graph.  DGs, not as much.

dkershaw10 replied:
Can we fold this into the very first table? And add a column to indicate whether this type of group can be created through MS Graph or not.

BTW - there's no way to tell a Yammer group other than it doesn't have Outlook conversations.  Similar thing for Teams, in that it has channels.

We should also **strongly** recommend that if developers are thinking about DG creation, they should instead use Office 365 Groups.

viv-liu replied:
I checked with Yammer. You actually can check conversations. It might be empty, but it exists. So I've found no way to see if a group is a Yammer or Outlook group unless I go to /beta and GET the group, and see "resourceProvisioningOptions" = "YammerProvisioning"

I can fold the mail-enabled SG into the top table, but I don't think we should fold Yammer groups up there. You can't even get the posts in a Yammer group in the API, nor tell that this is a Yammer group. I'll move Mail-enabled SG up though.

viv-liu commented on api-reference/v1.0/resources/groups_overview.md:
```diff
@@ -0,0 +1,126 @@
+Groups are collections of [users](https://developer.microsoft.com/en-us/graph/docs/api-reference/v1.0/resources/user) who share access to resources in Microsoft services or in your app. Microsoft Graph provides you with APIs to create and manage a variety of different types of groups and group functionality to suit your scenario needs. All operations in Microsoft Graph on groups require administr ...
+
+**NOTE**
+Groups are only supported for work or school accounts, and not supported for Microsoft personal accounts.
```
ok. personal MS accounts then

Commit:
Another iteration based on feedback

Commit:
Updated with Mike's feedback

Commit:
Small md tweak

Commit:
Small tweaks again
